src/Functions/snowflake.cpp

```diff
@@ -105,32 +101,46 @@ class FunctionSnowflakeToDateTime : public IFunction
         return std::make_shared<DataTypeDateTime>(timezone);
     }
 
-    ColumnPtr executeImpl(const ColumnsWithTypeAndName & arguments, const DataTypePtr &, size_t input_rows_count) const override
+    template <typename T>
+    bool executeNonConst(const IColumn & col_src, ColumnDateTime::Container & res_data, size_t input_rows_count) const
     {
-        const auto & src = arguments[0];
-        const auto & src_column = *src.column;
-
-        auto res_column = ColumnUInt32::create(input_rows_count);
-        auto & res_data = res_column->getData();
-
-        if (const auto * src_column_non_const = typeid_cast<const ColumnInt64 *>(&src_column))
+        if (const auto * col_src_non_const = typeid_cast<const T *>(&col_src))
         {
-            const auto & src_data = src_column_non_const->getData();
+            const auto & src_data = col_src_non_const->getData();
             for (size_t i = 0; i < input_rows_count; ++i)
-                res_data[i] = static_cast<UInt32>(
-                    ((src_data[i] >> time_shift) + snowflake_epoch) / 1000);
+                res_data[i] = static_cast<UInt32>((src_data[i] >> time_shift) / 1000);
+            return true;
         }
-        else if (const auto * src_column_const = typeid_cast<const ColumnConst *>(&src_column))
+        return false;
+    }
+
+    bool executeConst(const IColumn & col_src, ColumnDateTime::Container & res_data, size_t input_rows_count) const
+    {
+        if (const auto * col_src_const = typeid_cast<const ColumnConst *>(&col_src))
         {
-            Int64 src_val = src_column_const->getValue<Int64>();
+            Int64 src_val = col_src_const->getValue<Int64>();
```

Comment:
I think this [causes](https://pastila.nl/?001dd8aa/643fda160e29f55954fb727659f5f8a0#aWDe8VXciCutPbZwlrQ29A==) issues with number bigger than int64 max. We should check if it is int64 or uint64.